### ISSUE TITLE
Generalizing fitting algorithm to MPO-MPO multiplication

### ISSUE THREAD

**Issue by shinaoka** [ITensor/ITensorTDVP.jl#47]
Thank you for developing the nice library!

I am wondering how to generalize the fitting algorithm to MPO-MPO multiplication. I am slowly understanding the internal structure of this library. If I understand correctly, this could be straightforwardly achieved by generalizing `ProjMPOApply`. Is this correct? Because I have some use cases in my projects, I could contribute to it.
https://github.com/ITensor/ITensorTDVP.jl/blob/main/src/projmpo_apply.jl

**Comment by mtfishman** [comment]
Good question. Ultimately we plan to allow more general inputs to `tdvp` beyond MPS, for example tree tensor networks (#43), so it seems reasonable to allow `MPO` as well.

A minimal thing you can do is just convert your MPO to an MPS (there isn't any restriction on MPS to have only one site index per site), and see what breaks. At some point I tested this out for using TDVP to evolve a density matrix. I think it all worked after I generalized some of the projected MPO code to handle the "dangling" indices properly: https://github.com/ITensor/ITensors.jl/commit/0101936601643e5b12bc79a8c73c9535ceb550b9

**Comment by shinaoka** [comment]
Thank you for the reply.

> Ultimately, we plan to allow more general inputs to tdvp beyond MPS, for example tree tensor networks (https://github.com/ITensor/ITensorTDVP.jl/pull/43), so it seems reasonable to allow MPO as well.
That'd be great!

In the meantime, I have just made a quick dirty hack to ITensorTDVP.jl and ITensors.jl.

I had to extend `ITensors.replacebond!` to MPO, but the rest was smooth.
The notebook was tested for ITensors v0.3.20.

https://github.com/shinaoka/ITensorTDVP.jl/blob/contract_mpo_mpo/examples/contract_mpo_mpo.ipynb

**Comment by emstoudenmire** [comment]
Agreed with Matt's answer about mapping to the MPS type and back, and tweaking the code to handle doubled site indices. Sounds like it worked!

So do I understand from your code, literally the only change needed was defining replacebond!, and the rest just worked? That's great if so.

**Comment by shinaoka** [comment]
@emstoudenmire 

> the only change needed was defining replacebond!, and the rest just worked? 

Correct! Let me clean up the code.

**Comment by shinaoka** [comment]
Here is a small patch:
https://github.com/shinaoka/ITensorTDVP.jl/tree/contract_mpo_mpo_clean

Changing a few lines in `function ITensors.contract` was enough!

**Comment by emstoudenmire** [comment]
Yes a PR would be great. Thanks! And for now you could put the MPO version of `replacebond!` in the contract_mpo_mps.jl file and I'll work myself on a separate PR for ITensors.jl to generalize `replacebond!` to handle MPOs, and can remove that (nearly) duplicate code myself later.

**Comment by shinaoka** [comment]
I am a bit confused. Do we need the MPO version of `replacebond!`?
